### PR TITLE
Dev zone voting

### PR DESCRIPTION
**Pull request by Mopplikus**
Draft PR since this is still a WIP.

This adds the following changes:
- Implemented a display showing the player in which zone they currently are, at the top of the map. When the player is in none of the zones, it displays "None". If the location can't be retrieved, it displays "Unknown"
- If the player is in a zone, it displays either:
  - An attack button if the section the player is in does not own the current zone
  - A defend button if the section the player is in owns the zone
  - A timer until the next take-over, if there is no take-over right now or if the player already launched an attack
- Refactored the back-end behind a more convenient `FirebaseInterface` to have one single common point for all of the back-end operations to go through instead of a previous per-fragment basis; refactored the scoreboard fragment as such
- Added a constructor to the main activity to allow for injection during testing

This is still WIP since it raises a few questions, in particular how asynchronous operations are managed. I've implemented it using Futures here, but this will stall the main thread until the back-end gives an update, so we'll probably want a loading icon of some sort to be displayed while waiting. Another thing is determining if the player already launched an attack or not; for now, this is on a screen basis, meaning that it is reset once the fragment is closed and re-opened later, and should be fixed.

Also, for now, the main activity holds the firebase interface, and this will have to be changed if we can attach it to the application context instead.

Also, no tests have been added yet, especially due to the complexity of setting the testing environments in terms of position relative to a zone and time (take-over vs. other).

### CODE REVIEW

**Comment by Mopplikus** [comment]
Update: the tests have been implement, this feature is should be ready to be merged.

File: app/src/main/java/com/github/campus_capture/bootcamp/firebase/PlaceholderFirebaseInterface.java
Comment (Ant-One) [inline app/src/main/java/com/github/campus_capture/bootcamp/firebase/PlaceholderFirebaseInterface.java:23]: A bit biased your tests /s

File: app/src/main/java/com/github/campus_capture/bootcamp/fragments/MapsFragment.java
Comment (Ant-One) [inline app/src/main/java/com/github/campus_capture/bootcamp/fragments/MapsFragment.java:63]: Nice work with the futures

File: app/src/main/java/com/github/campus_capture/bootcamp/fragments/MapsFragment.java
Comment (Ant-One) [inline app/src/main/java/com/github/campus_capture/bootcamp/fragments/MapsFragment.java:102]: We will need to change the `.allowMainThreadQueries()` as it might slow down the UI if queries takes more time. But that's on my PR

**Review by Ant-One — COMMENTED** [review]
Not much to say expect nice work and nicely done. For now, it doesn't run because the firebase interface is not working (it crashes when trying to get who owns the zone). But according to your tests and to manual testing it should work.

Do we merge this now or do we wait for the firebase interaface merge? To be discussed tomorrow

File: app/src/main/java/com/github/campus_capture/bootcamp/firebase/PlaceholderFirebaseInterface.java
Comment (Mopplikus) [reply app/src/main/java/com/github/campus_capture/bootcamp/firebase/PlaceholderFirebaseInterface.java:23]: Coverage is coverage ¯\_(ツ)_/¯

**Comment by Mopplikus** [comment]
@Ant-One the bug which caused a crash when retrieving zone owners has been fixed, I simply forgot to correctly implement the placeholder. I also fixed some minor issues on the way. I'll let someone else do the final review.

File: .idea/androidTestResultsUserPreferences.xml
Comment (Blopgrop) [inline .idea/androidTestResultsUserPreferences.xml:1]: we should put this file in the git ignore a some point

File: app/src/main/java/com/github/campus_capture/bootcamp/fragments/MapsFragment.java
Comment (Blopgrop) [inline app/src/main/java/com/github/campus_capture/bootcamp/fragments/MapsFragment.java:272]: could use ternary operator to satisfy codeclimate btw

File: app/src/main/java/com/github/campus_capture/bootcamp/map/MapScheduler.java
Comment (Blopgrop) [inline app/src/main/java/com/github/campus_capture/bootcamp/map/MapScheduler.java:125]: Those are Runnables and  not functions cause they need to be used by an asynchron scheduler right ?

**Review by Blopgrop — APPROVED** [review]
I'm vibing way more with this new way of interfacing the firebase db ! It's more readable and scalable too so great plus. I remember you talked about potential de-encapsulation, as class that needs the interface will have access to all the querries that other classes can do, should we still take care of that or we'll survive with it ?
Other then that you did a nice implementation of the main feature of the game so good work !

File: .idea/androidTestResultsUserPreferences.xml
Comment (Mopplikus) [reply .idea/androidTestResultsUserPreferences.xml:1]: I'll do it now so that this issue is resolved once and for all

File: app/src/main/java/com/github/campus_capture/bootcamp/fragments/MapsFragment.java
Comment (Mopplikus) [reply app/src/main/java/com/github/campus_capture/bootcamp/fragments/MapsFragment.java:272]: Well spotted

File: app/src/main/java/com/github/campus_capture/bootcamp/map/MapScheduler.java
Comment (Mopplikus) [reply app/src/main/java/com/github/campus_capture/bootcamp/map/MapScheduler.java:125]: Correct! There are some other ways to solve this but this was mainly done to alleviate the methods, and since the tasks do not change it makes more sense to store them locally.

**Comment by Mopplikus** [comment]
@Blopgrop as you mentioned, having one unified interface does leak unrelated operations to all classes using them, but the benefit of centralising all the accesses through a single interface outweighs fragmenting throughout the codespace I believe. From an abstraction point of view this solution is not perfect, but I believe it is more intuitive like this, and anyone working on back-end implementations will have an easier time instead of scouring through each of the interfaces and will be grateful for this technical simplification.

Regarding your other comments, I'll push a quick update on the two little questions you raised, to have one final clean solution (after all, this task affects a _lot_ of how we'll be working from now on so it better be clean and without any afterthoughts)

**Pull request merged**
